buildkite/data_source_pipeline_template.go

```diff
@@ -0,0 +1,202 @@
+				"Unable to get pipeline template",
+				fmt.Sprintf("Error getting pipeline template: %s", err.Error()),
+			)
+			return
+		}
+
+		// Convert from Node to getNodeNodePipelineTemplate type
+		if pipelineTemplateNode, ok := apiResponse.GetNode().(*getNodeNodePipelineTemplate); ok {
+			if !ok {
+				resp.Diagnostics.AddError(
+					"Unable to get pipeline template",
+					"Error getting pipeline template: nil response",
+				)
+				return
+			}
+			updatePipelineTemplateDatasourceState(&state, *pipelineTemplateNode)
+		}
+	} else if !state.Name.IsNull() {
+		matchFound := false
+		err := retry.RetryContext(ctx, timeouts, func() *retry.RetryError {
+			var cursor *string
+			for {
+				r, err := getPipelineTemplates(
+					ctx,
+					pt.client.genqlient,
+					pt.client.organization,
+					cursor)
+				if err != nil {
+					if isRetryableError(err) {
+						return retry.RetryableError(err)
+					}
+					resp.Diagnostics.AddError(
+						"Unable to read pipeline templates",
+						fmt.Sprintf("Unable to read pipeline templates: %s", err.Error()),
+					)
+					return retry.NonRetryableError(err)
+				}
+
+				for _, template := range r.Organization.PipelineTemplates.Edges {
+					if template.Node.Name == state.Name.ValueString() {
+						matchFound = true
+						updatePipelineTemplateDatasourceFromNode(&state, template.Node)
+						break
+					}
+				}
+
+				// If no match found and at the last page, break
+				if matchFound || !r.Organization.PipelineTemplates.PageInfo.HasNextPage {
+					break
+				}
+
+				// Move to next cursor
+				cursor = &r.Organization.PipelineTemplates.PageInfo.EndCursor
+			}
+			return nil
+		})
+
+		if err != nil {
+			resp.Diagnostics.AddError("Unable to find pipeline template", err.Error())
+			return
+		}
+
+		if !matchFound {
+			resp.Diagnostics.AddError("Unable to find pipeline template",
+				fmt.Sprintf("Could not find a pipeline template with name \"%s\"", state.Name.ValueString()))
+			return
+		}
+	}
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+}
+
+func updatePipelineTemplateDatasourceState(ptds *pipelineTemplateDatasourceModel, ptn getNodeNodePipelineTemplate) {
+	ptds.ID = types.StringValue(ptn.Id)
+	ptds.UUID = types.StringValue(ptn.Uuid)
+	ptds.Available = types.BoolValue(ptn.Available)
+	ptds.Configuration = types.StringValue(ptn.Configuration)
+	ptds.Description = types.StringPointerValue(ptn.Description)
+	ptds.Name = types.StringValue(ptn.Name)
+}
+
+func updatePipelineTemplateDatasourceFromNode(ptds *pipelineTemplateDatasourceModel, ptn getPipelineTemplatesOrganizationPipelineTemplatesPipelineTemplateConnectionEdgesPipelineTemplateEdgeNodePipelineTemplate) {
+	ptds.ID = types.StringValue(ptn.Id)
+	ptds.UUID = types.StringValue(ptn.Uuid)
+	ptds.Available = types.BoolValue(ptn.Available)
+	ptds.Configuration = types.StringValue(ptn.Configuration)
+	ptds.Description = types.StringPointerValue(ptn.Description)
+	ptds.Name = types.StringValue(ptn.Name)
+}
```

Comment:
This could use generics further down the line as the input types contain the same fields, we could pass in a `[T Node]` which represents an interface that either of those types fulfil.

Reply:
Huh - I actually tried it with a generic first pass; I recall there was some typing errors with an implementation that way... one to look at soon 👍